Src/Compilers/Core/MSBuildTask/BuildClient.cs

```diff
@@ -0,0 +1,398 @@
+        /// The IsConnected property on named pipes does not detect when the client has disconnected
+        /// if we don't attempt any new I/O after the client disconnects. We start an async I/O here
+        /// which serves to check the pipe for disconnection. 
+        ///
+        /// This will return true if the pipe was disconnected.
+        /// </summary>
+        private static async Task CreateMonitorDisconnectTask(
+            NamedPipeClientStream pipeStream,
+            CancellationToken cancellationToken)
+        {
+            var buffer = new byte[0];
+
+            while (!cancellationToken.IsCancellationRequested && pipeStream.IsConnected)
+            {
+                // Wait a tenth of a second before trying again
+                await Task.Delay(100, cancellationToken).ConfigureAwait(false);
+
+                try
+                {
+                    Log("Before poking pipe.");
+                    await pipeStream.ReadAsync(buffer, 0, 0, cancellationToken).ConfigureAwait(false);
+                    Log("After poking pipe.");
+                }
+                // Ignore cancellation
+                catch (OperationCanceledException) { }
+                catch (Exception e)
+                {
+                    // It is okay for this call to fail.  Errors will be reflected in the 
+                    // IsConnected property which will be read on the next iteration of the 
+                    LogException(e, "Error poking pipe");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the file path of the executable that started this process.
+        /// </summary>
+        /// <param name="processHandle"></param>
+        /// <param name="flags">Should always be 0: Win32 path format.</param>
+        /// <param name="exeNameBuffer">Buffer for the name</param>
+        /// <param name="bufferSize">
+        /// Size of the buffer coming in, chars written coming out.
+        /// </param>
+        [DllImport("Kernel32.dll", EntryPoint = "QueryFullProcessImageNameW", CharSet = CharSet.Unicode)]
+        static extern bool QueryFullProcessImageName(
+            IntPtr processHandle,
+            int flags,
+            StringBuilder exeNameBuffer,
+            ref int bufferSize);
+
+        private const int MAX_PATH_SIZE = 260;
+
+        /// <summary>
+        /// Get all process IDs on the current machine that have executable names matching
+        /// the pipe name and started from the <paramref name="expectedPath" />.
+        /// </summary>
+        private static IEnumerable<int> GetAllProcessIds(string expectedPath)
+        {
+            // Get all the processes with the right base name.
+            Process[] allProcesses = Process.GetProcessesByName(PipeName);
+            Log("Found {0} existing processes with matching base name", allProcesses.Length);
+
+            foreach (Process process in allProcesses)
+            {
+                using (process)
```

Comment:
It looks like if the loop stops before processing all elements, the remaining `Process` objects will not be disposed.